### PR TITLE
Use min_specialization in the remaining rustc crates

### PR DESCRIPTION
**Pull request by matthewjasper**
This adds a lot of `transmute` calls to replace the unsound uses of specialization.
It's ugly, but at least it's honest about what's going on.

cc #71420, @RalfJung

### CODE REVIEW

**Comment by matthewjasper** [comment]
Is there anyone in @rust-lang/compiler willing to review this?

File: src/librustc_metadata/rmeta/encoder.rs [hunk 1]
```diff
@@ -228,8 +228,13 @@ impl<'tcx> SpecializedEncoder<LocalDefId> for EncodeContext<'tcx> {
     }
 }
 
-impl<'tcx> SpecializedEncoder<Ty<'tcx>> for EncodeContext<'tcx> {
-    fn specialized_encode(&mut self, ty: &Ty<'tcx>) -> Result<(), Self::Error> {
+impl<'a, 'b, 'tcx> SpecializedEncoder<&'a ty::TyS<'b>> for EncodeContext<'tcx>
```
Comment (oli-obk) [inline src/librustc_metadata/rmeta/encoder.rs:231]: are the two separate lifetimes `'a` and `'b` really necessary here? Couldn't they just be one and use `Ty<'a>`?

File: src/librustc_metadata/rmeta/encoder.rs [hunk 2]
```diff
@@ -228,8 +228,13 @@ impl<'tcx> SpecializedEncoder<LocalDefId> for EncodeContext<'tcx> {
...
+where
+    &'a ty::TyS<'b>: UseSpecializedEncodable,
+{
+    fn specialized_encode(&mut self, ty: &&'a ty::TyS<'b>) -> Result<(), Self::Error> {
+        debug_assert!(self.tcx.lift(ty).is_some());
```
Comment (oli-obk) [inline src/librustc_metadata/rmeta/encoder.rs:236]: Is there a measurable perf loss to just doing `lift` and unwrap and skipping the transmute?

File: src/librustc_metadata/rmeta/encoder.rs [hunk 3]
```diff
@@ -251,12 +256,18 @@ impl<'tcx> SpecializedEncoder<interpret::AllocId> for EncodeContext<'tcx> {
...
 
-impl<'tcx> SpecializedEncoder<&'tcx [(ty::Predicate<'tcx>, Span)]> for EncodeContext<'tcx> {
+impl<'a, 'b, 'tcx> SpecializedEncoder<&'a [(ty::Predicate<'b>, Span)]> for EncodeContext<'tcx> {
     fn specialized_encode(
         &mut self,
-        predicates: &&'tcx [(ty::Predicate<'tcx>, Span)],
+        predicates: &&'a [(ty::Predicate<'b>, Span)],
     ) -> Result<(), Self::Error> {
-        ty_codec::encode_spanned_predicates(self, predicates, |ecx| &mut ecx.predicate_shorthands)
+        debug_assert!(predicates.iter().all(|(p, _)| self.tcx.lift(p.kind()).is_some()));
```
Comment (oli-obk) [inline src/librustc_metadata/rmeta/encoder.rs:264]: I think you can just lift the entire borrowed slice instead of each element

**Comment by oli-obk** [comment]
> It's ugly, but at least it's honest about what's going on.

I agree.

r? @oli-obk

r=me with the comments resolved in some way

File: src/librustc_metadata/rmeta/encoder.rs (hunk 1 above)
Comment (matthewjasper) [reply src/librustc_metadata/rmeta/encoder.rs:231]: Unfortunately not. Specializing impls can't repeat type/lifetime parameters.

File: src/librustc_metadata/rmeta/encoder.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_metadata/rmeta/encoder.rs]: Ok, looks like there are some serious losses here. Let's benchmark against the transmute variant and if that has no perf losses, let's merge it.

**Comment by nikomatsakis** [comment]
One thing I want to mention:

The specialization hack that we use was **meant** to be replaced with a custom trait for serialization, but that was blocked on the ability to use custom derive -- now that we have custom derive, couldn't we make the encodable trait carry a `'tcx` explicitly? Perhaps things have changed too much to make that possible? (cc @eddyb)

Also, how much is any of this documented? I cast my eye over the diff but I admit I've always found the encoding system rather confusing.

**Comment by Aaron1011** [comment]
@bors r- r=oli-obk

This wasn't showing in the queue at https://buildbot2.rust-lang.org/homu/queue/rust

**Pull request merged**